metaflow/subprocess_manager.py

```diff
@@ -0,0 +1,278 @@
+import os
+import sys
+import time
+import signal
+import shutil
+import hashlib
+import asyncio
+import tempfile
+from typing import List
+
+
+def hash_command_invocation(command: List[str]):
+    concatenated_string = "".join(command)
+    current_time = str(time.time())
+    concatenated_string += current_time
+    hash_object = hashlib.sha256(concatenated_string.encode())
+    return hash_object.hexdigest()
+
+
+class LogReadTimeoutError(Exception):
+    pass
+
+
+class SubprocessManager(object):
+    def __init__(self):
+        self.commands = {}
+
+    async def __aenter__(self):
+        return self
+
+    async def __aexit__(self, exc_type, exc_value, traceback):
+        await self.cleanup()
+
+    async def run_command(self, command: List[str], env=None, cwd=None):
+        command_id = hash_command_invocation(command)
+        self.commands[command_id] = CommandManager(command, env, cwd)
+        await self.commands[command_id].run()
+        return command_id
+
+    def get(self, command_id: str) -> "CommandManager":
+        return self.commands.get(command_id, None)
+
+    async def cleanup(self):
+        for _, v in self.commands.items():
+            await v.cleanup()
+
+
+class CommandManager(object):
+    def __init__(self, command: List[str], env=None, cwd=None):
+        self.command = command
+
+        if env is None:
+            env = os.environ.copy()
+        self.env = env
+
+        if cwd is None:
+            cwd = os.getcwd()
+        self.cwd = cwd
+
+        self.process = None
+        self.run_called = False
+        self.log_files = {}
+
+        signal.signal(signal.SIGINT, self.handle_sigint)
```

Comment:
This would technically register this multiple times. I don't think this is supported or what you want. I am not sure how best to handle multiple commands though.

Reply:
let's say if you press CTRL+C on the main terminal while using the subprocess manager to run 2 commands -- both of them will terminate.

However, if you specifically do `kill -SIGINT PID` where PID is for any of the process we spawned out of the above two -- only that particular process will be killed.

This is the current behaviour. As we can see, only the process pertaining to stream B with alpha as 20 gets killed.

<img width="1457" alt="Screenshot 2024-04-17 at 2 26 11 AM" src="https://github.com/Netflix/metaflow/assets/155637867/e7f03672-1e63-4e4b-a3c7-ff0bcaa36133">

Reply:
resolving this, but can demo in the meeting...